src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -827,9 +827,11 @@ void ClientChannel::ExternalConnectivityWatcher::Notify(
   // automatically remove all watchers in that case.
   if (state != GRPC_CHANNEL_SHUTDOWN) {
     chand_->work_serializer_->Run(
-        [this]() ABSL_EXCLUSIVE_LOCKS_REQUIRED(chand_->work_serializer_) {
-          RemoveWatcherLocked();
-        },
+        [pThis = Ref()]()
```

Comment:
I don't know what's causing the bug you're seeing, but I don't see how this can possibly fix it.  The final ref to the `ExternalConnectivityWatcher` will not be removed until after `RemoveWatcherLocked()` runs, since the `ConnectivityStateTracker` is holding a ref to it, so there's no way we can be missing a ref here.

Reply:
The logs clear show that for watcher `0x7f528c0050e0 `, `ExternalConnectivityWatcher::Notify` was scheduled to run after the destructor `~ExternalConnectivityWatcher`.

I believe it happens because of the SHUTDOWN caused watchers being removed from the state tracker at 
https://github.com/grpc/grpc/blob/master/src/core/lib/transport/connectivity_state.cc#L173

Reply:
Why do you have a pending connectivity state watch when you're shutting down?  This sounds to me like a bug on the ruby side.  You should cancel any pending connectivity watch before shutting down the CQ.

Reply:
cc @apolcyn